[CBRD-23842] Support CDC interface

joohok opened the pull request:
http://jira.cubrid.org/browse/CBRD-23842

This PR is for CDC interface.

hgryoo committed:
[CBRD-23903] Fix slip: BIGINT return type of COUNT for hash scan (#2913)

http://jira.cubrid.org/browse/CBRD-23903

Because of this slip, in hash scanning, counting logic increments for an uninitialized value so that an unexpected value is returned at the end. Specifically, by the db_make_int(), value.data.i is initialized to 0. but the following counting logic increments from the uninitialized value.data.bigint value.

shparkcubrid committed:
[CBRD-24002] A problem where a predicate is pushed to a subquery that includes a limit clause. (#2914)

http://jira.cubrid.org/browse/CBRD-24002

fix to not push predicate into subquery having instnum in where clause.

shparkcubrid committed:
[CBRD-24009] Plan cache is not updated when view changes including inline view. #2915

http://jira.cubrid.org/browse/CBRD-24009

spec.entity_name is not needed to copy. This routine is added in CUBRIDSUS-13334.
but CUBRIDSUS-13334 related tests do not cause any problems now.
I delete the added routine in CUBRIDSUS-13334.

shparkcubrid committed:
[CBRD-24013] The problem that the order by + limit clause of subquery is deleted #2933

http://jira.cubrid.org/browse/CBRD-24013

fix to do not remove orderby+limit of subquery

shparkcubrid committed:
[CBRD-24010] Predicate push dose not work in nested subqueries. (#2921)

http://jira.cubrid.org/browse/CBRD-24010
http://jira.cubrid.org/browse/CBRD-24001

This change solves the problem below.
1. nested subquery problem (CBRD-24010)
2. indirect predicate problem (CBRD-24001)
3. Perform predicate push after changing outer join to inner join (newly discovered problem)

For resolving the CBRD-24010, I modified that mq_copypush_sargable_terms() is invoked in pre-order.
Since a predicate push is performed before a transitive predicate generation, it is occurred a problem of the CBRD-24001. For resloving the issue, I moved the location of mq_copypush_sargable_terms() in mq_optimize().

shparkcubrid committed:
[CBRD-24014] Problem that nullable predicate is not pushable. (#2934)

http://jira.cubrid.org/browse/CBRD-24014

fix that only in case of outer join spec, nullable predicate cannot be pushed.

Commit:
initial commit before sync

wrlawodms committed:
[CBRD-24016] It may fail to load Rootclass mop due to TDE (#2938)

http://jira.cubrid.org/browse/CBRD-24016

kisoo-han committed:
[CBRD-24021] Put libncurses6 symbolic link to CUBRID/lib as libncurses5 (#2955)

Commit:
apply review for supplement-dml(#2912)

joohok commented:
Sorry for commit 'apply review for supplement-dml(#2912)' that contains synchronize code. 
in order to test the code, I inserted some code.

Please review only two files 
- src/storage/heap_file.c
- src/storage/heap_file.h

I applied the reviews in https://github.com/CUBRID/cubrid/pull/2912 

thanks

Commit:
apply review for system parameter (#2908)

Commit:
apply review for log append function (#2911)

hornetmj committed:
[CBRD-23763] fix coredump that occurs when executing the copydb utility with -d option (#2950)

http://jira.cubrid.org/browse/CBRD-23763

* fix coredump that occurs when executing the copydb utility with -d option because it uses a deallocated global variable after calling tp_final function.
* tp_domain_free(...) -> area_free (tp_Domain_area, dom) -> assert (tp_Domain_area != NULL)
   - tp_Domain_area is null value because tp_final() has already called by boot_server_all_finalize() with BOOT_SHUTDOWN_ALL_MODULES shutdown mode.
   - So, I change the shutdown mode from BOOT_SHUTDOWN_ALL_MODULES to BOOT_SHUTDOWN_EXCEPT_COMMON_MODULES

wrlawodms commented on src/storage/heap_file.c:
```diff
@@ -881,6 +881,8 @@ STATIC_INLINE int heap_copy_chain (THREAD_ENTRY * thread_p, PAGE_PTR page_heap,
 STATIC_INLINE int heap_get_last_vpid (THREAD_ENTRY * thread_p, const HFID * hfid, VPID * last_vpid)
   __attribute__ ((ALWAYS_INLINE));
 
+STATIC_INLINE bool check_supplemental_log (THREAD_ENTRY * thread_p, OID * classoid) __attribute__ ((ALWAYS_INLINE));
```
seems to need a prefix.

shparkcubrid committed:
[CBRD-24018] Segment fault occurs when there are more than 33 analytic functions in the select list. (#2945)

http://jira.cubrid.org/browse/CBRD-24018

segment fault occured because of analytic function index overflow.
I add routines to check index of ANALYTIC_OPT_MAX_FUNCTIONS.

shparkcubrid committed:
[CBRD-24017] The problem that predicate push is not performed without changing the outer join to the inner join. (#2941)

http://jira.cubrid.org/browse/CBRD-24017

info_spec.appears, info->nullable is not initialized properly. and I fixed it to handle right outer join effectively using mq_is_outer_join_spec().

wrlawodms commented on src/storage/heap_file.c:
```diff
@@ -22473,6 +22705,25 @@ heap_insert_logical (THREAD_ENTRY * thread_p, HEAP_OPERATION_CONTEXT * context,
+    }
+  else
+    {
+      context->do_supplemental_log = false;
```
You should initialize this value in heap_clear_operation_context(), and in turn, you don't have to set false here.

wrlawodms replied:
It could be applied to other heap operations too.

wrlawodms commented on src/storage/heap_file.c:
```diff
@@ -22708,6 +22976,24 @@ heap_delete_logical (THREAD_ENTRY * thread_p, HEAP_OPERATION_CONTEXT * context)
+    {
+      context->do_supplemental_log = false;
+    }
+
```
How about creating a function for this repeated work?

wrlawodms commented on src/storage/heap_file.c:
```diff
@@ -22529,6 +22786,12 @@ heap_insert_logical (THREAD_ENTRY * thread_p, HEAP_OPERATION_CONTEXT * context,
+      if (context->do_supplemental_log && !heap_is_big_length (context->recdes_p->length))
+	{
+	  LSA_COPY (&context->supp_redo_lsa, &tdes->tail_lsa);
+	}
```
How about wrapping these into an inline function or macro?

wrlawodms commented on src/storage/heap_file.c:
```diff
@@ -22529,6 +22786,12 @@ heap_insert_logical (THREAD_ENTRY * thread_p, HEAP_OPERATION_CONTEXT * context,
 				context->recdes_p, is_mvcc_op, context->is_redistribute_insert_with_delid);
+
+      /* redo lsa for SUPPLEMENT_INSERT log */
+      if (context->do_supplemental_log && !heap_is_big_length (context->recdes_p->length))
```
I don't think `heap_is_big_length()` works as you expect. See the `heap_insert_handle_multipage_record()` in which `context->recdes_p->length` became sizeof OID.

wrlawodms commented on src/storage/heap_file.c:
```diff
@@ -22486,6 +22737,12 @@ heap_insert_logical (THREAD_ENTRY * thread_p, HEAP_OPERATION_CONTEXT * context,
       goto error;
     }
 
+/*SUPPLEMENT_INSERT REDO LSA, INSERT REC_BIGONE  */
+  if (context->do_supplemental_log && heap_is_big_length (context->recdes_p->length))
+    {
+      LSA_COPY (&context->supp_redo_lsa, &tdes->tail_lsa);
+    }
+
```
How about putting it into `heap_insert_handle_multipage_record()`?

wrlawodms commented on src/storage/heap_file.c:
```diff
@@ -21410,6 +21503,7 @@ heap_delete_home (THREAD_ENTRY * thread_p, HEAP_OPERATION_CONTEXT * context, boo
 	  rec_address.pgptr = context->home_page_watcher_p->pgptr;
 	  rec_address.vfid = &context->hfid.vfid;
 	  rec_address.offset = context->oid.slotid;
+
```
```suggestion
```

wrlawodms commented on src/storage/heap_file.c:
```diff
@@ -22486,6 +22737,12 @@ heap_insert_logical (THREAD_ENTRY * thread_p, HEAP_OPERATION_CONTEXT * context,
+/*SUPPLEMENT_INSERT REDO LSA, INSERT REC_BIGONE  */
+  if (context->do_supplemental_log && heap_is_big_length (context->recdes_p->length))
+    {
+      LSA_COPY (&context->supp_redo_lsa, &tdes->tail_lsa);
```
I recommend you using `logtb_find_current_tran_lsa()` instead of directly accessing `tdes->tail_lsa`. It also make you to be able to get rid of `tdes = LOG_FIND_CURRENT_TDES (thread_p);`s. This could be applied to all heap operations.

wrlawodms replied:
refer to `heap_log_delete_physical()`

wrlawodms commented on src/storage/heap_file.c:
```diff
@@ -21438,9 +21540,19 @@ heap_delete_home (THREAD_ENTRY * thread_p, HEAP_OPERATION_CONTEXT * context, boo
+	  /* log operation */
+	  heap_log_delete_physical (thread_p, context->home_page_watcher_p->pgptr, &context->hfid.vfid, &context->oid,
+				    &context->home_recdes, is_reusable, NULL);
+	}
```
```suggestion
	  /* log operation */
	  heap_log_delete_physical (thread_p, context->home_page_watcher_p->pgptr, &context->hfid.vfid, &context->oid,
				    &context->home_recdes, is_reusable, context->do_supplemental_log ? &context->supp_undo_lsa : NULL);
```

wrlawodms commented on src/storage/heap_file.c:
```diff
@@ -21141,8 +21211,18 @@ heap_delete_relocation (THREAD_ENTRY * thread_p, HEAP_OPERATION_CONTEXT * contex
+	{
+	  /* if reusable slot is deleted, then postponed log (RVHF_MARK_REUSABLE_SLOT) will appended last. 
+	   * So, current log lsa after heap_log_delete_physical can points unexpected log, other than delete log. */
+	  heap_log_delete_physical (thread_p, context->forward_page_watcher_p->pgptr, &context->hfid.vfid, &forward_oid,
```
same opinion to `heap_delete_home()`

wrlawodms commented on src/storage/heap_file.c:
```diff
@@ -4160,6 +4162,28 @@ heap_copy_chain (THREAD_ENTRY * thread_p, PAGE_PTR page_heap, HEAP_CHAIN * chain
+  else
+    {
+      return false;
+    }
```
```suggestion
  return prm_get_integer_value (PRM_ID_SUPPLEMENTAL_LOG) > 0 
             && !thread_p->no_supplemental_log
             && !oid_is_system_class (classoid));
```

Commit:
Add no-supplemental-log hint (#2909)

wrlawodms committed:
[CBRD-23987] Move slots at once when shifting slots up or down. (#2953)

http://jira.cubrid.org/browse/CBRD-23987

Move [start_slot_p, end_slot_p) to the end_slot_p direction at once when serveral slots are shifted up or down.

Commit:
server side client request handling (#2944)

Commit:
tested cubrid_log_connect and cubrid_log_find_lsa

Commit:
Interpret DML log record

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -321,6 +369,9 @@ static cubthread::daemon *log_Check_ha_delay_info_daemon = NULL;
 static cubthread::daemon *log_Flush_daemon = NULL;
 static std::atomic_bool log_Flush_has_been_requested = {false};
+
+static cubthread::daemon *cdc_log_Producer_daemon = NULL;
```
```suggestion
static cubthread::daemon *cdc_Log_producer_daemon = NULL;
```

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10408,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+  LOG_LSA lsa;
+
+  LOG_PAGE *log_page_p = NULL;
+  char log_pgbuf[IO_MAX_PAGE_SIZE + MAX_ALIGNMENT];
```
IO_MAX_PAGE_SIZE -> LOG_PAGESIZE?

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10408,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+      pthread_mutex_unlock (&cdc_Gl.lsa_init_lock);
+    }
+
+  temp_logbuf = (CDC_TEMP_LOGBUF *) malloc (2 * sizeof (CDC_TEMP_LOGBUF));
```
Where is this released? Why do you allocate it dynamically, not on the stack memory?

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10408,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+  CDC_LOGINFO_ENTRY *log_info_entry;
+
+  /*temp_logbuf consists of 2 entry, every time fetching new page, page will be stored in temp_logbuf[pageid%2] */
+  CDC_TEMP_LOGBUF *temp_logbuf = NULL;
```
One buffer always seems to store the page `log_page_p` refers to. Is it necessary? we can define a macro like `LOG_READ_DOES_FIT` macro.

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10408,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+
+  tmpbuf_index = process_lsa.pageid % 2;
+  temp_logbuf[tmpbuf_index].pageid = process_lsa.pageid;
+  memcpy (temp_logbuf[tmpbuf_index].log_page_p, log_page_p, IO_MAX_PAGE_SIZE);
```
They seem to be initialized always together. It can be a function or you can define an assignment operator of CDC_TEMP_LOGBUF for safety and simplicity.

hornetmj replied:
Plus, you can remove the memcpy() if you calculate this before calling the logpb_fetch_page() and pass one of the temp_logbuf to it.

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10408,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+  temp_logbuf[tmpbuf_index].pageid = process_lsa.pageid;
+  memcpy (temp_logbuf[tmpbuf_index].log_page_p, log_page_p, IO_MAX_PAGE_SIZE);
+
+  log_info_entry = (CDC_LOGINFO_ENTRY *) malloc (sizeof (CDC_LOGINFO_ENTRY));
```
Where is this released? Why do you allocate it dynamically, not on the stack memory?

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10408,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+
+	  LSA_COPY (&cdc_Gl.last_produced_lsa, &cur_log_rec_lsa);
+
+	  log_info_entry = (CDC_LOGINFO_ENTRY *) malloc (sizeof (CDC_LOGINFO_ENTRY));
```
You seem to allocate it above, aren't you?

wrlawodms replied:
And, if it weren't, ` if (log_info_entry->length != 0)` above looks cause a segfault.

Commit:
fixes to handle zipped supplemental logs

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -170,6 +170,33 @@ struct archive_log_header_scan_context
 
+/*It will be moved to the CDC module and will define overflow.h in that module  */
+
+struct overflow_first_part
```
I'd like to ask the reason for defining it again. It's because this is used on the client-side?

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -170,6 +170,33 @@ struct archive_log_header_scan_context
+typedef struct overflow_first_part OVERFLOW_FIRST_PART;
+
+/* *INDENT-OFF* */
+static std::unordered_map <TRANID, char *> cdc_tran_user; /*to clear when log producer ends suddenly */
```
```suggestion
static std::unordered_map <TRANID, char *> cdc_Tran_user; /*to clear when log producer ends suddenly */
```

wrlawodms replied:
Why is it a global variable? It seems that it is used only in `cdc_log_producer()`.

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10418,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+	      return NO_ERROR;
+	    }
+	}
+      else if (log_rec_hdr->type == LOG_ABORT && log_rec_hdr->trid == trid)
```
What if the trx is committed? It seems that it searches to the end of the log file.

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10418,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+		  /* can not find user. It meets abort log. So, ignore the logs from this transaction */
+		  tran_ignore.insert (std::make_pair (trid, 1));
+		}
+	    }
```
LOG_COMMIT or LOG_ABORT isn't the end of log of the trx? You can search a user name after either record?

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10418,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+	      is_unzip_supplement = log_unzip (supp_zip, supplement_length, supplement_data);
+	    }
+
+	  if (is_zip_supplement && is_unzip_supplement)
```
`is_zip_supplement ` means the log record is zipped, and `is_unzip_supplement` means that it succeeds to unzip. If it is right, the name of variables should be distinguishable.

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10418,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+	    {
+	      supp_zip = log_append_get_zip_undo (thread_p);
+
+	      is_unzip_supplement = log_unzip (supp_zip, supplement_length, supplement_data);
```
Needs error handling when log_unzip() fails

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10418,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+		}
+
+	      supplement_data = supp_zip->log_data;
+	    }
```
This seems to be able to put into the upper if-condition.

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10418,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+		  else
+		    {
+		      /* can not find user. It meets abort log. So, ignore the logs from this transaction */
+		      tran_ignore.insert (std::make_pair (trid, 1));
```
After this, the sppl. log which is being processed seems to has to be skipped.

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10418,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+
+	      if (cdc_get_recdes (thread_p, temp_logbuf, NULL, NULL, &redo_lsa, &redo_recdes) != NO_ERROR)
+		{
+		  break;
```
Just skip this record?

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10418,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+		  break;
+		}
+
+	      cdc_make_dml_loginfo (thread_p, trid, tran_user, INSERT, classoid, NULL, &redo_recdes, log_info_entry);
```
error handling

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10418,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+	      memcpy (&undo_lsa, supplement_data + sizeof (OID), sizeof (LOG_LSA));
+	      memcpy (&redo_lsa, supplement_data + sizeof (OID) + sizeof (LOG_LSA), sizeof (LOG_LSA));
+
+	      cdc_get_recdes (thread_p, temp_logbuf, &undo_lsa, &undo_recdes, &redo_lsa, &redo_recdes);
```
undo_recdes->data can be malloced in `cdc_get_recdes()`, but I can't find `free()` for it.

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10418,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+
+      switch (log_rec_hdr->type)
+	{
+	case LOG_SUPPLEMENTAL_INFO:
```
Question: Is this only for LOG_SUPPLEMENT_UNDO_RECORD?

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10418,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+      log_rec_hdr = LOG_GET_LOG_RECORD_HEADER (log_page_p, &process_lsa);
+
+      LOG_READ_ADD_ALIGN (thread_p, sizeof (*log_rec_hdr), &process_lsa, log_page_p);
+
```
Is there no case in which we should traverse back to find a undo/redo data except for log records for overflow file?

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10418,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+		    undo_zip_ptr = log_zip_alloc (IO_PAGESIZE);
+		    if (undo_zip_ptr == NULL)
+		      {
+			return ER_OUT_OF_VIRTUAL_MEMORY;
```
```suggestion
			return er_errid();
```

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10418,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+			goto end;
+		      }
+
+		    logpb_copy_from_log (thread_p, undo_data, undo_length, &process_lsa, log_page_p);
```
In this function, log_page_p can go forward several pages. We don't need to consider it for `temp_logbuf`?

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10418,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+	    break;
+	  }
+	case LOG_MVCC_DIFF_UNDOREDO_DATA:
+	case LOG_MVCC_UNDOREDO_DATA:
```
Could you tell me the reason you make this routine on your own, not using `log_get_undo_record()`?

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10418,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+		  {
+		    goto end;
+		  }
+	      }
+
```
```suggestion
	      }
	      else 
	      {
	       assert (false);
	      }

```

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10418,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+	    redo_length = mvcc_undoredo->undoredo.rlength;
+	    undo_length = mvcc_undoredo->undoredo.ulength;
+
+	    if (LOG_IS_DIFF_UNDOREDO_TYPE (log_rec_hdr->type) == true)
+	      {
+		is_diff = true;
+	      }
```
```suggestion
		is_diff = (OG_IS_DIFF_UNDOREDO_TYPE (log_rec_hdr->type);
```

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10418,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+
+		if (process_lsa.offset + redo_length < (int) LOGAREA_SIZE)
+		  {
+		    redo_data = (char *) (log_page_p->area + process_lsa.offset);
```
Is it OK no to skip the undo part?

wrlawodms replied:
Even if the undo part can be skipped, how about checking the checking and skipping. This code assumes the undo part of RVHF_MVCC_INSERT is 0, but it may be changed later.

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10418,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+    }
+
+/* Get REDO RECDES from redo lsa */
+  if (redo_lsa != NULL)
```
Have you considered the generalized way to extract a redo log record like log_recovery_redo()? They don't deal with log records rcvindex by rcvindex. I guess you see that code already. please tell me why you should do it case by case?

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10418,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+	{
+	  heap_value = &attr_info.values[i];
+	  oldval_deforder = heap_value->read_attrepr->def_order;
+	  memcpy (&old_values[oldval_deforder], &heap_value->dbvalue, sizeof (DB_VALUE));
```
How about just using old_attr_info and new_attr_info ?

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10418,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+      has_pk = 0;
+    }
+
+  if (record_length > 0)
```
can be <= 0?

wrlawodms commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10418,3430 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+		}
+	    }
+
+	  ptr = or_pack_int (ptr, num_cond_col);
```
not num_change_col ?

wrlawodms reviewed (commented):
I took a look at it in a hurry because of the deadline. Please re-request review later when it's ready.

Commit:
fix heap_logical_insert to handle overflow record

Commit:
manage queue and thread

hornetmj commented on src/communication/network_interface_sr.c:
```diff
@@ -10240,7 +10240,7 @@ scdc_get_lsa (THREAD_ENTRY * thread_p, unsigned int rid, char *request, int reql
   ptr = or_unpack_int64 (request, &input_time);
 
-  error = cdc_get_lsa (thread_p, input_time, &start_lsa);
+  error = cdc_get_lsa (thread_p, &input_time, &start_lsa);
```
The un-used variable 'b_start_lsa' should be removed.

hornetmj commented on src/parser/parse_tree.h:
```diff
@@ -1199,8 +1199,9 @@ typedef enum
-    /* SELECT key information from index b-tree instead of table record data */
+  PT_HINT_SELECT_KEY_INFO = 0x80000000,	/* 1000 0000 0000 0000 0000 0000 0000 0000 */
+  /* SELECT key information from index b-tree instead of table record data */
+  PT_HINT_NO_SUPPLEMENTAL_LOG = 0x100000000
```
Adds comment in the same manner. And I think you need to leave the purpose of this hint and descriptions.

hornetmj commented on src/api/cubrid_log.c:
```diff
@@ -0,0 +1,1533 @@
+      return CUBRID_LOG_INVALID_FUNC_CALL_STAGE;
+    }
+
+  if ((classoid_arr == NULL && arr_size != 0) || (classoid_arr != NULL && arr_size == 0))
```
```suggestion
  if (classoid_arr == NULL || arr_size == 0)
```

hornetmj commented on src/api/cubrid_log.c:
```diff
@@ -0,0 +1,1533 @@
+
+  if ((classoid_arr == NULL && arr_size != 0) || (classoid_arr != NULL && arr_size == 0))
+    {
+      return CUBRID_LOG_INVALID_CLASSOID_ARR_SIZE;
```
If classoid_arr == NULL, then the error code is not correct. So, it needs to be reconsidered.

hornetmj commented on src/api/cubrid_log.c:
```diff
@@ -0,0 +1,1533 @@
+      return CUBRID_LOG_INVALID_FUNC_CALL_STAGE;
+    }
+
+  if ((user_arr == NULL && arr_size != 0) || (user_arr != NULL && arr_size == 0))
```
```suggestion
  if (user_arr == NULL || arr_size == 0)
```

hornetmj commented on src/api/cubrid_log.c:
```diff
@@ -0,0 +1,1533 @@
+#include "object_representation.h"
+
+#define CUBRID_LOG_ERROR_HANDLING(e, v) \
+  (err_code) = (e); \
```
check the alignment.

hornetmj commented on src/api/cubrid_log.c:
```diff
@@ -0,0 +1,1533 @@
+
+  if (curr_size + tracelog_size > g_trace_log_filesize)
+    {
+      fseek (g_trace_log, 0, SEEK_SET);
```
In this case, the loggings wrote previously exist in the trace log file. It needs to handle.

hornetmj commented on src/api/cubrid_log.c:
```diff
@@ -0,0 +1,1533 @@
+      CUBRID_LOG_ERROR_HANDLING (CUBRID_LOG_INVALID_ID, trace_errbuf);
+    }
+
+  if (password == NULL)
```
The default 'PUBLIC' and 'DBA' users have no passwords. Do you have any idea how to handle it?

hornetmj commented on src/api/cubrid_log.c:
```diff
@@ -0,0 +1,1533 @@
+      CUBRID_LOG_ERROR_HANDLING (CUBRID_LOG_INVALID_FUNC_CALL_STAGE, trace_errbuf);
+    }
+
+  if (timestamp <= 0)
```
The timestamp variable is a pointer type. So, this condition check is invalid.

Anyway, I think there is no reason that the timestamp variable is a pointer type. How about reconsider changing it to timestamp type?

hornetmj commented on src/api/cubrid_log.c:
```diff
@@ -0,0 +1,1533 @@
+
+cubrid_log_error:
+
+  if (recv_data != NULL && recv_data != reply)
+    {
+      free_and_init (recv_data);
+    }
+
+  queue_entry = css_find_queue_entry (g_conn_entry->buffer_queue, rid);
+  if (queue_entry != NULL)
+    {
+      queue_entry->buffer = NULL;
+      css_queue_remove_header_entry_ptr (&g_conn_entry->buffer_queue, queue_entry);
+    }
```
This routine is repeated in several functions. So, I think it seems nice to make a macro or inline function.

hornetmj commented on src/api/cubrid_log.c:
```diff
@@ -0,0 +1,1533 @@
+
+  ptr = or_unpack_int (recv_data, &reply_code);
+
+  if (reply_code == NO_ERROR || reply_code == ER_CDC_LSA_NOT_FOUND)
```
Isn't the ER_CDC_LSA_NOT_FOUND error? Does it have the LSA and timestamp information?

If it is, Is it OK to return just CUBRID_LOG_SUCCESS?

hornetmj commented on src/api/cubrid_log.c:
```diff
@@ -0,0 +1,1533 @@
+      CUBRID_LOG_ERROR_HANDLING (CUBRID_LOG_INVALID_OUT_PARAM, trace_errbuf);
+    }
+
+  memcpy (&g_next_lsa, lsa, sizeof (LOG_LSA));
```
The 'g_next_lsa' and 'lsa' should be equals in the normal processing. So, this condition should be checked together when the function argument 'lsa' is checked.

hornetmj commented on src/api/cubrid_log.c:
```diff
@@ -0,0 +1,1533 @@
+
+  rc = cubrid_log_extract_internal (&g_next_lsa, &num_infos, &total_length);
+
+  if (rc != CUBRID_LOG_SUCCESS)
```
The return value 'CUBRID_LOG_SUCCESS_WITH_NO_LOGITEM' should be considered.

hornetmj commented on src/api/cubrid_log.c:
```diff
@@ -0,0 +1,1533 @@
+
+  if (dml->num_changed_column)
+    {
+      // now, just malloc for validation and later, optimize it.
```
I think it is unnecessary to malloc/free every time. So, I request you to make a simple data buffer. (after releasing the second test version)

hornetmj commented on src/api/cubrid_log.c:
```diff
@@ -0,0 +1,1533 @@
+      CUBRID_LOG_ERROR_HANDLING (CUBRID_LOG_INVALID_FUNC_CALL_STAGE, trace_errbuf);
+    }
+
+  if (log_item_list == NULL)
```
The address of the log_item_list is equaled to g_log_items. So, It needs to check.

hornetmj reviewed (commented):
I reviewed the client(API + lib) first.

hornetmj commented on src/communication/network.h:
```diff
@@ -248,6 +248,14 @@ enum net_server_request
 
   NET_SERVER_VACUUM_DUMP,
+
+  /*FOR CDC */
```
```suggestion
  /* For CDC */
```

hornetmj replied:
Or remove this comment.

hornetmj commented on src/communication/network_interface_sr.h:
```diff
@@ -185,6 +185,7 @@ extern void srepl_set_info (THREAD_ENTRY * thread_p, unsigned int rid, char *req
 extern void srepl_log_get_append_lsa (THREAD_ENTRY * thread_p, unsigned int rid, char *request, int reqlen);
 extern void slocator_check_fk_validity (THREAD_ENTRY * thread_p, unsigned int rid, char *request, int reqlen);
 extern void slogwr_get_log_pages (THREAD_ENTRY * thread_p, unsigned int rid, char *request, int reqlen);
+
```
```suggestion
```

hornetmj commented on src/communication/network_interface_sr.h:
```diff
@@ -238,4 +239,11 @@ extern void sloaddb_destroy (THREAD_ENTRY * thread_p, unsigned int rid, char *re
 extern void sloaddb_update_stats (THREAD_ENTRY * thread_p, unsigned int rid, char *request, int reqlen);
 extern void ssession_stop_attached_threads (void *session);
+
+/*for CDC*/
```
```suggestion
/* For CDC */
```

hornetmj replied:
Or remove this comment.

hornetmj commented on src/communication/network_interface_sr.c:
```diff
@@ -8019,6 +8020,8 @@ slogwr_get_log_pages (THREAD_ENTRY * thread_p, unsigned int rid, char *request,
   return;
 }
 
+
+
```
```suggestion
```

hornetmj commented on src/communication/network_interface_sr.c:
```diff
@@ -10223,3 +10226,209 @@ ssession_stop_attached_threads (void *session)
+
+      for (int i = 0; i < num_extraction_user; i++)
+	{
+	  ptr = or_unpack_string (ptr, &dummy_user);
```
```suggestion
	  ptr = or_unpack_string_nocopy (ptr, &dummy_user);
```

hornetmj commented on src/communication/network_interface_sr.c:
```diff
@@ -10223,3 +10226,209 @@ ssession_stop_attached_threads (void *session)
+
+	  if (extraction_user[i] == NULL)
+	    {
+	      for (int j = 0; j <= i; j++)
```
```suggestion
	      for (int j = 0; j < i; j++)
```

hornetmj commented on src/communication/network_interface_sr.c:
```diff
@@ -10223,3 +10226,209 @@ ssession_stop_attached_threads (void *session)
+	      goto end;
+	    }
+
+	  memcpy (extraction_user[i], dummy_user, extraction_user_len);
```
How about use strncpy?

hornetmj commented on src/communication/network_interface_sr.c:
```diff
@@ -10223,3 +10226,209 @@ ssession_stop_attached_threads (void *session)
+      extraction_classoids = (uint64_t *) malloc (sizeof (uint64_t) * num_extraction_class);
+      if (extraction_classoids == NULL)
+	{
+	  if (num_extraction_user > 0)
+	    {
+	      for (int i = 0; i < num_extraction_user; i++)
+		{
+		  free (extraction_user[i]);
+		}
+
+	      free (extraction_user);
+	    }
```
```suggestion
```

hornetmj commented on src/communication/network_interface_sr.c:
```diff
@@ -10223,3 +10226,209 @@ ssession_stop_attached_threads (void *session)
+	  error = ER_OUT_OF_VIRTUAL_MEMORY;
+	  goto end;
+	}
+      for (int i = 0; i < num_extraction_class; i++)
```
```suggestion

      for (int i = 0; i < num_extraction_class; i++)
```

hornetmj commented on src/communication/network_interface_sr.c:
```diff
@@ -10223,3 +10226,209 @@ ssession_stop_attached_threads (void *session)
+    }
+
+  error =
+    cdc_set_configuration (max_log_item, timeout, all_in_cond, extraction_user, num_extraction_user,
```
```suggestion
    cdc_set_configuration (max_log_item, extraction_timeout, all_in_cond, extraction_user, num_extraction_user,
```

hornetmj replied:
And you might remove the unused variable 'timeout'.

hornetmj commented on src/communication/network_interface_sr.c:
```diff
@@ -10223,3 +10226,209 @@ ssession_stop_attached_threads (void *session)
+
+end:
+
+  ptr = or_pack_int (reply, error);
```
How about using the error_code iso error? I think it is consistent compare to the other function in this file.

hornetmj commented on src/communication/network_interface_sr.c:
```diff
@@ -10223,3 +10226,209 @@ ssession_stop_attached_threads (void *session)
+
+  ptr = or_pack_int (reply, error);
+  ptr = or_pack_log_lsa (ptr, &start_lsa);
+  or_pack_int64 (ptr, input_time);
```
My guess is that input_time is semantically incorrect. Therefore, it is recommended to modify the variable name of input_time according to the meaning.

hornetmj replied:
Do we need to send additional information(start_lsa, imput_time) in case of an error?

hornetmj reviewed (commented):
I reviewed the system parameter and network communications. But I will review the scdc_get_logitem_info and scdc_get_logitem functions again when I review the cdc_ function series because I need to understand how to work the cdc_Server_comm.is_sent flag.

hornetmj commented on src/api/cubrid_log.h:
```diff
@@ -0,0 +1,151 @@
+  extern int cubrid_log_set_extraction_user (char **user_arr, int arr_size);
+
+/* API for the preparation step */
+  extern int cubrid_log_connect_server (char *host, int port, char *dbname, char *id, char *password);
```
```suggestion
  extern int cubrid_log_connect_server (char *host, int port, char *dbname, char *user, char *password);
```

Commit:
DDL : supplemental log and interpretation

shparkcubrid committed:
[CBRD-24022] The problem of using reserved words of HQ as data filters in hierarchical queries (#2963)

http://jira.cubrid.org/browse/CBRD-24022

Exclude predicates with reserved words for HQ from data filter.

Commit:
fixed queue mechanizm and partition redistribution bug

hornetmj committed:
[CBRD-23994] fix coredump when calling vacuum_get_log_blockid() if vacuum_disable is set to true. (#2947)

http://jira.cubrid.org/browse/CBRD-23994

* fix coredump
* if the hidden system parameter 'vacuum_disable' is set to true, then the coredump is occurred whenever calling the vacuum_get_log_blockid() function. Because the vacuum_Data.log_block_npages is always set to 0 and it make a 'divide by 0' problem'

hornetmj committed:
[CBRD-24027] The 'er_log_vacuum' becomes the hidden system parameter. (#2970)

https://jira.cubrid.org/browse/CBRD-24027

kisoo-han committed:
[CBRD-24024] Change CMS submodule for gethapplyinfo (#2973)

Provide active_eof_lsa, copied_active_eof_lsa  to the clients

Commit:
fix thread synchronize, and server hanging when client is terminated abnormaly

Commit:
modify truncate from DML to DDL

hornetmj commented on src/transaction/log_impl.h:
```diff
@@ -759,6 +767,100 @@ typedef struct log_logging_stat
+
+typedef struct cdc_server_comm
+{
+  char *log_items;		/* log item list. it is used as buffer to send to client */
```
Do you want to call it log_item? I remembered you said it log_info.

hornetmj commented on src/transaction/log_impl.h:
```diff
@@ -788,6 +890,14 @@ extern char log_Name_volinfo[];
+/* *INDENT-OFF* */
+extern lockfree::circular_queue<CDC_LOGINFO_ENTRY *> *cdc_logInfo_queue;
+/* *INDENT-ON* */
+extern CDC_SERVER_COMM cdc_Server_comm;
```
How about combining the global structures related to the CDC?

I think it is not good that the mutex related to the cdc_logInfo_queue exists in other global variable cdc_Gl.init_queue_lock.

hornetmj replied:
How about initializing the members of a global variable at the same time?
(e.g. cdc_Gl.init_queue_lock is initialized in the cdc_log_producer_daemon_init() when booting but, others in the cdc_set_configuration())

IMHO, I don't think good readable code even though it works well.

hornetmj commented on src/transaction/log_impl.h:
```diff
@@ -759,6 +767,100 @@ typedef struct log_logging_stat
+  pthread_cond_t queue_init_cond;
+
+  uint64_t queue_size;
+  bool is_queue_initialized;
```
How about replacing to cdc_logInfo_queue != NULL check?

hornetmj commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10426,3742 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+      LSA_SET_NULL (&cdc_Gl.last_consumed_lsa);
+
+      pthread_cond_signal (&cdc_Gl.queue_init_cond);
+      pthread_mutex_unlock (&cdc_Gl.init_queue_lock);
```
```suggestion
      pthread_mutex_unlock (&cdc_Gl.init_queue_lock);
      pthread_cond_signal (&cdc_Gl.queue_init_cond);
```

hornetmj commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10426,3742 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+ */
+
+static int
+get_start_point_from_file (THREAD_ENTRY * thread_p, int arv_num, LOG_LSA * ret_lsa, time_t * time)
```
prefix.

hornetmj commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10426,3742 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+
+  while (!LSA_ISNULL (&process_lsa))
+    {
+      log_rec_header = LOG_GET_LOG_RECORD_HEADER (log_pgptr, &process_lsa);
```
At the first loop, the process_lsa.offset is 0.

I think that using the LOG_PAGE.hdr.offset is more correct.

hornetmj commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10426,3742 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+	}
+    }
+
+  if (is_found)
```
is_found is always true at this point.

hornetmj commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10426,3742 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+  else
+    {
+      /* NO ERROR */
+      if (active_start_time != 0 && active_start_time <= *time)
```
if active_start_time == *time, then just use ret_lsa as start_lsa.

And I think it is nice if a user could set the time as -1 (or something) then, the server returns a LSA from default action(e.g. the first LSA of the active log volume). It is my suggestion for usability.

hornetmj commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10426,3742 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+ */
+
+static int
+get_lsa_with_start_point (THREAD_ENTRY * thread_p, time_t * time, LOG_LSA * start_lsa)
```
prefix.

hornetmj commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10426,3742 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+
+  int error = NO_ERROR;
+
+  if (LSA_ISNULL (start_lsa))
+    {
+      is_active = true;
+    }
```
```suggestion
```

hornetmj replied:
It may be the assertion case.

hornetmj commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10426,3742 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+	  if (donetime->at_time >= *time)
+	    {
+	      *time = donetime->at_time;
+	      LSA_COPY (start_lsa, &log_rec_header->forw_lsa);
```
```suggestion
	      LSA_COPY (start_lsa, &forw_lsa);
```

hornetmj commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10426,3742 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+
+  LOG_REC_DONETIME *donetime;
+  LOG_REC_HA_SERVER_STATE *dummy;
+  time_t at_time;
```
```suggestion
```

hornetmj commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10426,3742 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+      if (fileio_is_volume_exist (arv_name) == true)
+	{
+	  vdes = fileio_mount (thread_p, log_Db_fullname, arv_name, LOG_DBLOG_ARCHIVE_VOLID, false, false);
+	  if (vdes != NULL_VOLDES)
```
error handling if vdes == NULL_VOLDES.

hornetmj commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10426,3742 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+	}
+      pthread_mutex_lock (&cdc_Gl.next_lsa_lock);
+      pthread_cond_signal (&cdc_Gl.lsa_init_cond);
+      pthread_mutex_unlock (&cdc_Gl.next_lsa_lock);
```
```suggestion
      pthread_cond_signal (&cdc_Gl.lsa_init_cond);
```

hornetmj commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10426,3742 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+  _er_log_debug (ARG_FILE_LINE, "cdc_daemons_init ");
+#endif
+  cdc_log_producer_daemon_init ();
+  cdc_thread_checker_daemon_init ();
```
Is this daemon thread absolutely necessary?

hornetmj commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10426,3742 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+    {
+      pthread_mutex_lock (&cdc_Gl.next_lsa_lock);
+      pthread_cond_signal (&cdc_Gl.nxio_lsa_cond);
+      pthread_mutex_unlock (&cdc_Gl.next_lsa_lock);
```
```suggestion
      pthread_cond_signal (&cdc_Gl.nxio_lsa_cond);
```

hornetmj commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10426,3742 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+	{
+	  pthread_mutex_unlock (&cdc_Gl.next_lsa_lock);
+	  goto end;
+	}
```
I think this condition never occurs.

hornetmj commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10426,3742 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+    }
+  pthread_mutex_unlock (&cdc_Gl.queue_consume_lock);
+
+  pthread_mutex_lock (&cdc_Gl.is_finalize_lock);
```
The previous 'if (cdc_Gl.is_finalize == true)' checking doesn't get the 'cdc_Gl.is_finalize_lock' mutex.

Anyway, I think we should reduce unnecessary condition variables and mutexes.

Commit:
support timestamp data type, and fix memory corruption

hornetmj commented on src/transaction/log_manager.c:
```diff
@@ -10214,6 +10426,3742 @@ logtb_tran_update_stats_online_index_rb (THREAD_ENTRY * thread_p, void *data, vo
+	case LOG_ABORT:
+	  LOG_REC_DONETIME * donetime;
+
+	  LOG_READ_ADVANCE_WHEN_DOESNT_FIT (thread_p, sizeof (*donetime), &process_lsa, log_page_p);
+	  if ((process_lsa.pageid % 2) != tmpbuf_index)
+	    {
+	      tmpbuf_index = (tmpbuf_index + 1) % 2;
+	      temp_logbuf[tmpbuf_index].pageid = process_lsa.pageid;
+	      memcpy (temp_logbuf[tmpbuf_index].log_page_p, log_page_p, IO_MAX_PAGE_SIZE);
+	    }
```
How about introducing a new macro that covers this execution similar to the LOG_READ_ADVANCE_WHEN_DOESNT_FIT macro. It is nice if we can remove memcpy function call.

shparkcubrid committed:
[CBRD-24028] Segment fault occurs when connections increase rapidly

http://jira.cubrid.org/browse/CBRD-24028

When the number of inputs becomes larger than the size of m_stack, m_ext_block is used. But it does not copy the old value of m_stack into m_ext_block. This means there are garbage values.
I fix to copy data from m_stack to m_ext_block at first extension.

Commit:
add supplemental log dump function to test

Commit:
supplemental log for system class is appended before server is booted

Commit:
ignore system class when dml processing

kisoo-han committed:
[CBRD-23862] Reset yylineno explicitly before start parsing (#2962)

Correction of incorrect error line number in csql

joohok committed:
Modify the condition when the object represents the class (#2981)

Commit:
free the not null object  when finalize

Commit:
fixed windows build error

Commit:
hanging in making ddl loginfo when user filter returns -1

Commit:
remove hanging when stop server

Commit:
fixed trigger supplemental log

Commit:
Merge branch 'feature/cdc-support' of https://www.github.com/cubrid/cubrid into feature/cdc-support

Commit:
Merge branch 'feature/cdc-support' into cdc-merged-version

beyondykk9 committed:
[CBRD-23956] If an error occurs when executing a MERGE query in Autocommit, a -550 (Unknown savepoint name UmsP_ #) error also occurs. (#2984)

http://jira.cubrid.org/browse/CBRD-23956

Commit:
support analyzing overflow data

wrlawodms committed:
[CBRD-23622] Truncate Table Performance Enhancements (merge) (#2861)

http://jira.cubrid.org/browse/CBRD-23622

TRUNCATE TABLE statement is used to delete all the data from an existing table but not the table itself. It has the same effect as DELETE FROM [table_name] semantically. 
In CUBRID, TRUNCATE is implemented internally using the DELETE FROM statement, which makes it scan and delete all the data record by record.
With this PR, it's done by deallocating pages that contain records.

It includes some details:
- Some specification changes and a new options, CASCADE.
- Still using DELETE when a table is refenced by an object reference.
- Some optimization releated to reacreation indexes during TRUNCATE.

hornetmj committed:
[CBRD-24030] The db_get_ha_server_state API function is expected to return the current ha state of the sever. (#2987)

https://jira.cubrid.org/browse/CBRD-24030

hornetmj committed:
[CBRD-24032] The --compress(-z) option of backupdb utility is made to default option. (#2988)

https://jira.cubrid.org/browse/CBRD-24032

Purpose
The --compress(-z) option of backupdb utility is made to default option.

Implementation
--no-compress option is introduced newly.
--compress(-z) option is deprecated and is removed from the usages.

Remarks
--no-compress option works correctly.
--compress(-z) option works correctly. (no error, backward compatibility)
The usage doesn't show the description of the --compress(-z) option.
The backupdb utility always compresses the backup data unless --no-compress option is used.

Commit:
fixes extraction timeout bug

Commit:
Merge branch 'develop' into feature/cdc-support

Commit:
Merge branch 'feature/cdc-support' of https://www.github.com/cubrid/cubrid into feature/cdc-support

Commit:
merge due to recent changes on ha_server_status and truncate

Commit:
modify naming that causes collision with other header files when builds in Windows

Commit:
get rid of unused hint

Commit:
fix log producer hanging when nxio_lsa is null

Commit:
add serial statement, and fixes bugs

kisoo-han committed:
[CBRD-23862] Reset yyline only when csql interactive (#2995)

Call csql_yyset_lineno (1); in executables/csql.c only when it is interactive mode. (not from file)

hornetmj committed:
[CBRD-24037] The LOG_DUMMY_HA_SERVER_STATE is logged even though the ha_mode is disabled. (#2996)

http://jira.cubrid.org/browse/CBRD-24037

Commit:
code refactoring for thread management

Commit:
refactoring when get undo record

Commit:
thread and queue management in scdc_find_lsa, cdc_loginfo_producer

Commit:
add debug log for cdc

Commit:
implemented thread synchronize, cdc_logging_debug and fixed memory bugs

Commit:
fixed bugs related to extracting overflow data and memory corruption due to log_append_get_zip_*

Commit:
fix indent and cppcheck error

Commit:
Merge remote-tracking branch 'upstream/develop' into feature/cdc-support

Commit:
Merge branch 'feature/cdc-support' into cdc-merged-version

Commit:
turn on ha_dummy_status when supplemental log is on

Commit:
fix cppcheck error

joohok commented:
code style failure is not from this PR.
